File: AlgoPersonalStudy/src/backjoon/silver/BJ_1012.java
package backjoon.silver;

import java.util.*;
import java.io.*;

public class BJ_1012 {

    static int[] delta_x = {0,1,0,-1}, delta_y = {1,0,-1,0};
    public static void main(String[] args) throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine());
        for(int t=0;t<T;t++){
            StringTokenizer st = new StringTokenizer(br.readLine());
            int M = Integer.parseInt(st.nextToken());
            int N = Integer.parseInt(st.nextToken());
            int K = Integer.parseInt(st.nextToken());

            int[][] map = new int[N][M];
            boolean[][] visited = new boolean[N][M];

            for(int k=0;k<K;k++){
                st = new StringTokenizer(br.readLine());
                int x = Integer.parseInt(st.nextToken());
                int y = Integer.parseInt(st.nextToken());
                map[y][x] = 1;
            }

            Queue<int[]> queue = new LinkedList<>();
            int answer = 0;

            for(int n=0;n<N;n++){
                for(int m=0;m<M;m++){
                    if(map[n][m] == 1){
                        if(!visited[n][m]){
                            visited[n][m] = true;
                            answer++;
                            queue.add(new int[]{n,m});
                            while(!queue.isEmpty()){
                                int[] tmp = queue.poll();
                                for(int d=0;d<4;d++){
                                    int dy = tmp[0] + delta_y[d];
                                    int dx = tmp[1] + delta_x[d];
                                    if(dy >= 0 && dy < N && dx >=0 && dx < M){
                                        if(map[dy][dx] == 1 && !visited[dy][dx]){
                                            visited[dy][dx] = true;
                                            queue.add(new int[]{dy,dx});
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
            System.out.println(answer);
        }
    }
}

File: AlgoPersonalStudy/src/_kotlin/backjoon/silver/BJ_1012.kt
package _kotlin.backjoon.silver

import java.io.*
import java.util.*

class BJ_1012 {
    val deltaX = arrayOf(0,1,0,-1)
    val deltaY = arrayOf(1,0,-1,0)
    fun solution(){
        val br = BufferedReader(InputStreamReader(System.`in`))
        val t = Integer.parseInt(br.readLine())

        repeat(t){
            val (m,n,k) = br.readLine().split(" ").map { it.toInt() }
            val map = Array(n) { Array(m) { 0 } }
            val visited = Array(n) {Array(m){false} }
            repeat(k) {
                val (x, y) = br.readLine().split(" ").map { it.toInt() }
                map[y][x] = 1
            }
            val queue = LinkedList<Array<Int>>()
            var answer = 0

            for(y:Int in 0 until n){
                for(x:Int in 0 until m){
                    if(map[y][x] == 1){
                        if(!visited[y][x]){
                            visited[y][x] = true
                            answer++
                            queue.add(arrayOf(y,x))
                            while(!queue.isEmpty()){
                                val tmp = queue.poll()
                                for(d:Int in 0 until 4){
                                    val dy = tmp[0] + deltaY[d]
                                    val dx = tmp[1] + deltaX[d]
                                    if (dy in 0..<n && dx in 0 ..< m) {
                                        if (map[dy][dx] == 1 && !visited[dy][dx]) {
                                            visited[dy][dx] = true
                                            queue.add(arrayOf(dy, dx))
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
            println(answer)
        }
    }
}
fun main(){
    val solutionClass = BJ_1012()
    solutionClass.solution()
}